### DISCUSSION TITLE
How to install libs install after build?

### DISCUSSION THREAD

**Discussion by Gounlaf** [dlemstra/Magick.Native#30] [Q&A] [answered]
Hi,
I don't understand how to install libraries one compiled.

What I understand is there are 2 packaging method (nuget and npm).
I would like to install lib in same folder as other (/usr/local/lib or something like this).

I tried to install via `cmake install`  (found cmake file in Q8 or Q16 folder), but I don't understand what it does really.

Any help would be welcome.

Thanks.

**Comment by dlemstra** [comment]
Could you tell me more about your end goal? I am a bit confused why you would want to install this library? This library is now bundled with Magick.NET and magick-wasm and never used independently.

**Reply by Gounlaf** [reply]
I would like to use it in a Kotlin/Java environment, making bindings on it.

I'm trying to replicate some features of Magick.NET in a JVM.

Following this tutorial, it works for ImageMagick : https://kotlinlang.org/docs/native-c-interop.html

I'm just trying to do it with Magick.Native.

**Reply by dlemstra** [reply]
Interesting... I was also considering to create something that could be used in Kotlin/Java but realized that would probably take me too much time to focus on now. You should copy the `.so` files from the `Q8/Q16/Q16-HDRI` folders and use them in your project. We could also maybe in the future create something that you could us but I don't know what that should be now.

**Reply by Gounlaf** [reply]
:ok_hand: I will try copying the `.so`.  and let you know if it works as expected.

Right now, I'm just trying to "replicate" Magick.NET (only the features I need).
Digging & digging, I found that you use Magick.Native as a "wrapper" if I understand it correctly?

Porting/Mirroring Magick.NET on the JVM is my final goal 😅 (all credits to you, Magick.NET is great).

**Reply by Gounlaf** [reply]
Hi @dlemstra,

What I have done so far:

- compile Magick.Native (and ImageMagick) within a Dockerfile
- extract .so/.a/.la files
- extract headers files

It compiles ; still don't know if it works 😅

**Reply by dlemstra** [reply, marked as answer]
You should only need the `libMagick.Native-*.dll.so` files. I have also recently added devcontainers that can be used to create the output files so you could use that instead of a Dockerfile. 

Maybe for now you should build them yourself but I think that it would be better to use the binaries that I provide in the future. Maybe a maven or gradle package? I don't think we should add this now but once you have something ready maybe we should add a maven or gradle package that you could use.?

And I would advise you to start with something simple like getting the `Features` property of the `MagickNET` class working. This is also what I did with the `magick-wasm` library that I created recently.

And of course feel free to reach out if you have any follow up questions or ideas.

**Reply by Gounlaf** [reply]
C universe is new for me 😅

What I understand is it's mandatory to provide the headers files. The interop library is generated with the headers files.
And the library (shared or static) is only necessary for the compilation/binary generation.

> I have also recently added devcontainers that can be used to create the output files so you could use that instead of a Dockerfile.

I've reused the "devcontainer" Dockerfile, just adding the compilation step for the Native library found in the github workflow file.

> Maybe for now you should build them yourself but I think that it would be better to use the binaries that I provide in the future. Maybe a maven or gradle package? I don't think we should add this now but once you have something ready maybe we should add a maven or gradle package that you could use.?

Of course! Right now, I'm just trying to have a poc working.

> And I would advise you to start with something simple like getting the Features property of the MagickNET class working.

Thanks for the advise :ok_hand: 

> And of course feel free to reach out if you have any follow up questions or ideas.

:+1:

**Reply by Gounlaf** [reply]
> And I would advise you to start with something simple like getting the Features property of the MagickNET class working.

I obtained this :tada: :

`Channel-masks(64-bit) Cipher `
